### PR TITLE
B-18155-Update-SIT-service-item-addresss-on-shipment-address-update

### PR DESCRIPTION
**Pull request by traskowskycaci**
## Summary

This work will update the destination SIT service item's final address when the prime changes the shipment address.

Of the two endpoints called out in the description of the story: createSITAddressUpdateRequest "Creates an update request for a SIT service item's final delivery address. A newly created update request is assigned the status 'REQUESTED' if the change in address is > 50 miles and automatically approved otherwise." I didn't touch that endpoint as I believe there is some ambiguity we need to discuss on the approval process for shipment address changes and service item address changes. There is also a future story to address the logic for changing the status of a service item, B-18033.

The other endpoint in the description was updateMTOShipment. Instead of this, I made changes to updateMTOShipmentAddress. When the prime changes the shipment address, the destination SIT service item's destination address id will also change and will match the shipment. There is no logic for the TOO approval process as part of this work.

- [x] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [x] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

To test, log in to the office app with the prime simulator role. Change the address of the shipment and verify in the database that the shipment and the SIT service items all share the same final destination address ID.

To test with the prime API, use the updateMTOShipmentAddress endpoint and provide the shipment ID, address ID that is being updated, and a request body with the address information, such as
{
  "id": "c56a4180-65aa-42ec-a945-5fd21dec0538",
  "streetAddress1": "123 Main Ave",
  "streetAddress2": "Apartment 9000",
  "streetAddress3": "Montmârtre",
  "city": "Anytown",
  "state": "AL",
  "postalCode": "90210",
  "country": "USA"
}

you can comment out the etag validation in mto_shipment_address_updater.go if you need to. When you get a successful response, again verify in the database that the addresses for the shipment and the SIT destination service items match what you provided as the address information.

### CODE REVIEW

File: pkg/services/mto_shipment/mto_shipment_address_updater.go [hunk 1]
```diff
@@ -112,6 +148,11 @@ func (f mtoShipmentAddressUpdater) UpdateMTOShipmentAddress(appCtx appcontext.Ap
 		return nil, apperror.NewQueryError("Address", err, "")
 	}
 
+	_, err = UpdateSITServiceItemDestinationAddressToMTOShipmentAddress(&mtoShipment.MTOServiceItems, newAddress, appCtx)
+	if err != nil {
+		return nil, apperror.NewQueryError("No updated service items on shipmetn address change", err, "")
```
Comment (danieljordan-caci) [inline pkg/services/mto_shipment/mto_shipment_address_updater.go:153]: typo

File: pkg/services/mto_shipment/mto_shipment_address_updater.go [hunk 2]
```diff
@@ -43,6 +43,42 @@ func isAddressOnShipment(address *models.Address, mtoShipment *models.MTOShipmen
 	return false
 }
 
+func UpdateSITServiceItemDestinationAddressToMTOShipmentAddress(mtoServiceItems *models.MTOServiceItems, newAddress *models.Address, appCtx appcontext.AppContext) (*models.MTOServiceItems, error) {
+	// Change the address ID of destination SIT service items to match the address of the shipment address ID
+	serviceItems := mtoServiceItems
+	var updatedMtoServiceItems models.MTOServiceItems
+	for _, s := range *serviceItems {
+		serviceItem := s
```
Comment (danieljordan-caci) [inline pkg/services/mto_shipment/mto_shipment_address_updater.go:51]: Is there any reason why you're making a temp variable like you do here? It's totally fine, just curious.

Comment (traskowskycaci) [reply pkg/services/mto_shipment/mto_shipment_address_updater.go:51]: Had a reason at the time but can't remember what it was lol. Cleaning that up now!

**Review by danieljordan-caci — APPROVED** [review]

File: pkg/services/mto_shipment/mto_shipment_address_updater.go (hunk 2 above)
Comment (traskowskycaci) [reply pkg/services/mto_shipment/mto_shipment_address_updater.go:50]: Got rid of the temp serviceItems variable but keeping the reassignment of s inside the loop to get around an "Implicit memory aliasing in for loop" error I was getting.

https://stackoverflow.com/questions/62446118/implicit-memory-aliasing-in-for-loop

**Review by TevinAdams — APPROVED** [review]
Tested, and service items were updated with the shipments updated address.

**Pull request merged**